blob/service.go

```diff
@@ -351,123 +362,187 @@ func (s *Service) retrieve(
+	for i, row := range header.DAH.RowRoots[inclusiveNamespaceStartRowIndex:] {
+		if namespace.IsOutsideRange(row, row) {
+			exclusiveNamespaceEndRowIndex = inclusiveNamespaceStartRowIndex + i
 			break
 		}
 	}
 
-	getCtx, getSharesSpan := tracer.Start(ctx, "get-shares-by-namespace")
+	// calculate the square size
+	squareSize := len(header.DAH.RowRoots) / 2
 
-	// collect shares for the requested namespace
-	namespacedShares, err := s.shareGetter.GetSharesByNamespace(getCtx, header, namespace)
-	if err != nil {
-		if errors.Is(err, share.ErrNotFound) {
-			err = ErrBlobNotFound
+	// get all the shares of the rows containing the namespace
+	_, getSharesSpan := tracer.Start(ctx, "get-all-shares-in-namespace")
+	// store the ODS shares of the rows containing the blob
+	rowsShares := make([]share.Share, 0, (exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)*squareSize)
+	// store the EDS shares of the rows containing the blob
+	rowsWithParityShares := make([][]shares.Share, exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)
+	for rowIndex := inclusiveNamespaceStartRowIndex; rowIndex < exclusiveNamespaceEndRowIndex; rowIndex++ {
+		for colIndex := 0; colIndex < squareSize*2; colIndex++ {
+			share, err := s.shareGetter.GetShare(ctx, header, rowIndex, colIndex)
+			if err != nil {
+				return nil, nil, err
+			}
+			if colIndex < squareSize {
+				rowsShares = append(rowsShares, share)
+			}
+			appShare, err := toAppShares(share)
+			if err != nil {
+				return nil, nil, err
+			}
+			rowsWithParityShares[rowIndex-inclusiveNamespaceStartRowIndex] = append(
+				rowsWithParityShares[rowIndex-inclusiveNamespaceStartRowIndex],
+				appShare[0],
+			)
 		}
-		getSharesSpan.SetStatus(codes.Error, err.Error())
-		return nil, nil, err
 	}
 
 	getSharesSpan.SetStatus(codes.Ok, "")
 	getSharesSpan.AddEvent("received shares", trace.WithAttributes(
-		attribute.Int64("eds-size", int64(len(header.DAH.RowRoots)))))
+		attribute.Int64("eds-size", int64(squareSize*2))))
 
-	var (
-		appShares = make([]shares.Share, 0)
-		proofs    = make(Proof, 0)
-	)
-
-	for _, row := range namespacedShares {
-		if len(row.Shares) == 0 {
-			// the above condition means that we've faced with an Absence Proof.
-			// This Proof proves that the namespace was not found in the DAH, so
-			// we can return `ErrBlobNotFound`.
-			return nil, nil, ErrBlobNotFound
+	// go over the shares until finding the requested blobs
+	for currentShareIndex := 0; currentShareIndex < len(rowsShares); {
+		currentShareApp, err := shares.NewShare(rowsShares[currentShareIndex])
+		if err != nil {
+			return nil, nil, err
 		}
 
-		appShares, err = toAppShares(row.Shares...)
+		// skip if it's a padding share
+		isPadding, err := currentShareApp.IsPadding()
 		if err != nil {
 			return nil, nil, err
 		}
+		if isPadding {
+			currentShareIndex++
+			continue
+		}
+		isCompactShare, err := currentShareApp.IsCompactShare()
```

Comment:
Is this something new? as I'm not aware about `compactShares`?

Reply:
those are transaction shares: https://celestiaorg.github.io/celestia-app/specs/shares.html#transaction-shares
Could be any Celestia transaction.